### PR TITLE
Add recursive argument to spack develop

### PR DESCRIPTION
**Pull request by psakievich**
This effort allows for a recursive develop call
which will traverse from the develop spec given back to the root(s) and mark all packages along the path as develop.

If people are doing development across the graph then paying fetch and full rebuild costs every time `spack install` is called is unnecessary and expensive. Since this requires an already concrete environment the exact versions can be used and there is no need to guess or provide versions for concrete specs.

Additional changes:

- Remove the constraint of only one develop spec.  Users can specify more than one unless they pass spec specific configurations like `--path` or `--build-directory`.
- Change the default check for the version to make a non-concrete CLI spec concrete.  The order now is:
   1. If the environment is concrete take the version in the already concretized environment (highest version, if there happens to be repeats)
   2. Next take `max(package_class.versions)`

### CODE REVIEW

**Comment by tgamblin** [comment]
Can you split this into two PRs -- one for the removal of the constraint and the other for the recursive option?

**Comment by psakievich** [comment]
I put them together because the recursive makes sense to default to just the package name since it is already concretized. So we should do remove constraint first and then recursive dependent on that merging IMO.

**Comment by psakievich** [comment]
@tgamblin see #46911. I preemptively added you as a reviewer 🙂

**Comment by psakievich** [comment]
I started a unit-test for this, but it's looking like a big lift at the moment. It needs a concretized environment to work, and so far all I've found is mock concretization of a spec. Seems like a notable gap in our testing infrastructure but would dramatically extend the scope of this PR to implement and should be a PR in and of itself IMHO.

**Comment by psakievich** [comment]
Testing this locally for now:

``` console
$ spack env activate --temp 
$ spack add exawind
$ spack concretize
$ spack develop -r trilinos
# trilinos actually has a fetch error that I've documented but never fixed #25370 
$ spack develop -r trilinos
# clones see yaml file below
```

``` yaml
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs:
  - exawind
  view: true
  concretizer:
    unify: true
  develop:
    nalu-wind:
      spec: nalu-wind@=master
    exawind:
      spec: exawind@=master
    trilinos:
      spec: trilinos@=16.0.0
```

**Comment by psakievich** [comment]
@haampie this should meet your `spack develop` needs as outlined on slack.

File: lib/spack/spack/cmd/develop.py [hunk 1]
```diff
@@ -123,6 +131,25 @@ def develop(parser, args):
...
+    if args.recursive:
+        concrete_specs = env.all_matching_specs(spec)
+        if not concrete_specs:
+            tty.msg(
+                "No matching specs found in the environment. "
+                "Recursive develop requires a concretized environment"
+            )
+        else:
+            for s in concrete_specs:
+                for parent in s.traverse_edges(direction="parents", root=True):
```
Comment (haampie) [inline lib/spack/spack/cmd/develop.py]: Use spack.traverse.traverse_nodes(concrete_specs) instead of a double loop and key by dag hash to uniquely visit each node. Your current code visits nodes repeatedly.

I don't think it applies here, but you have to be very careful with upwards traversal, since sometimes it follows edges from parents that are no descendants of any of concrete_specs. That happens when specs are taken from the db (where all in-edges of the entire db are added).

File: lib/spack/spack/cmd/develop.py [hunk 2]
```diff
@@ -123,6 +131,25 @@ def develop(parser, args):
...
+
+                    parent_args = copy(args)
+                    parent_args.spec = parent.spec.format("{name}@{version}")
+                    parent_args.recursive = False
+
+                    tty.debug(f"Recursive develop for {parent_args.spec}")
+                    develop(parser, parent_args)
```
Comment (haampie) [inline lib/spack/spack/cmd/develop.py]: Make the develop logic a function of its own that does not depend on argparse.

File: lib/spack/spack/cmd/develop.py [hunk 3]
```diff
@@ -123,6 +131,25 @@ def develop(parser, args):
...
+            tty.msg(
+                "No matching specs found in the environment. "
+                "Recursive develop requires a concretized environment"
+            )
+        else:
+            for s in concrete_specs:
+                for parent in s.traverse_edges(direction="parents", root=True):
+
+                    parent_args = copy(args)
+                    parent_args.spec = parent.spec.format("{name}@{version}")
```
Comment (haampie) [inline lib/spack/spack/cmd/develop.py]: You want to use dag hashes... name@version is not specific enough.

**Comment by haampie** [comment]
> @haampie this should meet your `spack develop` needs as outlined on slack.

I doesn't: what I meant was I want be able to select an individual spec from a concrete env, so that the following:

Screenshot 2024-10-19 at 10 31 01

would take the matching curl@8.4.0 and add `dev_path=...`, and update the hashes of dependents.

I don't want to run `spack -e . concretize -f` cause that takes forever, and risks getting a completely different spec.

Also, apparently `spack -e . concretize -f` is broken if I put curl in the develop section with root spec cmake:

Screenshot 2024-10-19 at 10 34 20

So, in short: I wanna be able to add `dev_path=...` post-hoc, to a single spec selected from a concrete environment, without re-concretization.

**Comment by psakievich** [comment]
> So, in short: I wanna be able to add dev_path=... post-hoc, to a single spec selected from a concrete environment, without re-concretization.

Ah well I can add the single spec logic to this PR (prefer concrete if a version is omitted), but me thinks skipping re-concretization is a more fundamental change for another PR.

**Comment by psakievich** [comment]
Thinking about this more I think to skip concretization we'd likely need to change how develop is recognized (make it an environment property rather than a spec property). It messes with a lot of fundamental stuff though. The other option I can think of is a surgical change to the DAG. Modify specs, change hashes, but don't actually solve.

**Comment by haampie** [comment]
Note that `spack spec cmake ^curl dev_path=/tmp` works ;) With that in mind the whole `develop: ...` section is kind of redundant to start with...

> The other option I can think of is a surgical change to the DAG. Modify specs, change hashes, but don't actually solve.

Yeah, this is my suggestion too. In my opinion a major flaw in the develop command is that you don't know what to develop until the env is concretized, yet it wants to fetch sources ahead of time (including resources and patches which depend on variants, compiler choices). The order should've been: concretize first, then mark things as dev spec, then fetch sources.

**Comment by psakievich** [comment]
@haampie That order makes sense to me. The only reason I see develop needs to be part of the spec with a variant is to mark for differentiation of the install/future reuse. From an install point dev spec doesn't really matter? I guess it does in that you probably don't want to reuse a develop spec or rather don't want it masquerading as something it is not. Other than that I think it should be an environment property and not a spec property with the order you outlined. Oh and hash computation. New variant ensures new hash. 

I do like that develop can happen pre concretization for reuse of developer environments etc, but I don't see why that would have to change. 

I suspect that the install situation could be alleviated by not installing develop specs or their dependents into the same install tree. Have it be a special one. That actually aligns well with another request of having develop specs link and possibly rebuild out of a build tree and skip installation all together. Rational being linking to build directly means tighter changes (update a header and not have to reinstall just rebuild) and less file system work.

File: lib/spack/spack/cmd/develop.py (hunk 3 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: We can't use dag hashes b/c they will change when we concretize. I went generic to allow hitting multiple instances. I have always tried to go most generic for dev specs out of principle. I could do the `spec.short_spec` but if someone has a development environment with repeat specs (+cuda/~cuda) then `spack develop` can't support both.

File: lib/spack/spack/cmd/develop.py (hunk 1 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: Do you think there is a better traversal logic I could get to hit the intended intersection of the DAG? I read the warnings about upward traversal, and I do find it concerning.  I could see a visit each node in the DAG and check a condition if there is an accessible and sufficient condition on the specs to say "I depened on spec foo in my subgraph?".


Maybe something like:
``` python
if f"^{dev_spec.name}" in parent_spec.fullname:
   develop(parent_spec)
```

**Comment by psakievich** [comment]
I made a few more updates.  I refactored so that the core of developing a spec is 3 steps:

1) confirm it is concrete and pick a default if it is not (default to highest version in environment if concrete env, otherwise max version)
2) setup the source code
3) update the config/spack.yaml

These steps should be taken regardless of if a spec is from recursive collection, reading and updating the current develop specs or just CLI specified specs.

So those differentiating cases were moved to a generator function. I also removed the constraint of only being able to list one spec as develop at the CLI, but preserve that if `--path` or `--build-directory` are passed.  Someday I would like those to be projections so then they could be spec aware/valid for the multi spec case.

**Comment by psakievich** [comment]
I also can't seem to figure out why this unit-test keeps failing. I have tested updating a develop spec from the CLI and it works properly.

File: lib/spack/spack/cmd/develop.py [hunk 4]
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+                " --build-directory arguments"
+            )
+        else:
+            for spec in specs:
+                if args.recursive:
+                    concrete_specs = env.all_matching_specs(spec)
+                    if not concrete_specs:
+                        tty.msg(
+                            "No matching specs found in the environment. "
+                            "Recursive develop requires a concretized environment"
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]: This needs to say which spec didn't match, and should make clear that this spec is being skipped but others will still be developed.

File: lib/spack/spack/cmd/develop.py [hunk 5]
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
 
-    # If user does not specify --path, we choose to create a directory in the
-    # active environment's directory, named after the spec
-    path = args.path or spec.name
-    if not os.path.isabs(path):
-        abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
-    else:
-        abspath = path
+    entry = {"spec": str(spec)}
+    if path != spec.name:
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]:
```suggestion
    if path and path != spec.name:
```

File: lib/spack/spack/cmd/develop.py [hunk 6]
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+                f"packages:{spec.name}:package_attributes:build_directory:{build_dir}",
                 env.scope_name,
             )
-        _update_config(spec, path)
+        # add develop spec and update path
+        if specified_path:
+            # always overwrite path if it was supplied
+            _update_config(spec, specified_path)
+        else:
+            _update_config(spec, spec.name)
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]: If the path is just defaulting to the spec name, we don't need to pass it in at all
```suggestion
        # always overwrite path if it was supplied
        _update_config(spec, specified_path)
```

File: lib/spack/spack/cmd/develop.py [hunk 7]
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+    spec: spack.spec.Spec,
+    specified_path: Optional[str] = None,
+    build_dir: Optional[str] = None,
+):
+    """
+    Update the spack.yaml file with additions or changes from a develop call
+    """
+    tty.debug(f"Updating develop config for {env.name} transactionally")
+
+    # this check has to happen outside the write transaction otherwise updates don't seem to happen
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]: Can you fill me in a bit on what was happening with this? I don't understand why it would be necessary and that concerns me.

File: lib/spack/spack/cmd/develop.py [hunk 8]
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
-        if os.path.exists(abspath):
-            if args.force:
-                shutil.rmtree(abspath)
-            else:
-                msg = "Path %s already exists and cannot be cloned to." % abspath
-                msg += " Use `spack develop -f` to overwrite."
-                raise SpackError(msg)
 
-        _retrieve_develop_source(spec, abspath)
+def update_env_file(
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]: I think this would be easier to read as `update_env` since it's interacting with the `Environment` through the normal `spack.config` interface, not the `EnvironmentManifestFile` interface

File: lib/spack/spack/cmd/develop.py [hunk 9]
```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
-                msg += " because its path already exists."
-                tty.msg(msg)
-                continue
+def assure_concrete_spec(env: spack.environment.Environment, spec: spack.spec.Spec):
+    version = spec.versions.concrete_range_as_version
+    if not version:
+        # first check environment for a matching concrete spec
+        matching_specs = env.all_matching_specs(spec)
+        if matching_specs:
+            version = matching_specs[0].version
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py:90]: This seems unsafe -- won't it cause problems for any matching spec other than the highest version?

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/develop.py (hunk 9 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py:90]: Yes, but that is the consequence of only allowing one develop spec per package.  Perhaps we should just throw in this instance?

File: lib/spack/spack/cmd/develop.py (hunk 7 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: This was just for unit-tests to pass. It worked from the CLI.  I think it had to do with process forks?  I'm not sure exactly. I spend a lot of time trying to debug but all my debugging instances were getting hung up in the subprocesses and the inspection/print statements looked like the values were correct, but the test always failed for updating the develop spec.

File: lib/spack/spack/cmd/develop.py (hunk 9 above)
Comment (becker33) [reply lib/spack/spack/cmd/develop.py:90]: Yeah I think if we won't properly handle multiple specs with different versions, we should raise if we hit that.

File: lib/spack/spack/cmd/develop.py [hunk 10]
```diff
@@ -84,86 +79,152 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+                " --build-directory arguments"
+            )
+        else:
+            for spec in specs:
+                if args.recursive:
+                    concrete_specs = env.all_matching_specs(spec)
+                    if not concrete_specs:
+                        tty.warn(
+                            f"{spec.name} has no matching concrete specs in the environment and "
+                            "will be skipped. Recursive develop requires a concretized"
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]:
```suggestion
                            "will be skipped. `spack develop --recursive` requires a concretized"
```

File: lib/spack/spack/cmd/develop.py (hunk 9 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py:90]: That's the new behavior and I added a test for it.

**Comment by psakievich** [comment]
@spackbot fix style

File: lib/spack/spack/cmd/develop.py [hunk 11]
```diff
@@ -84,86 +79,152 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+        matching_specs = env.all_matching_specs(spec)
+        if matching_specs:
+            version = matching_specs[0].version
+            test_spec = spack.spec.Spec(f"{spec}@{version}")
+            for m_spec in matching_specs:
+                if not test_spec.satisfies(m_spec):
+                    raise SpackError(
+                        f"{spec.name}: has conflicting specs and only one develop spec per package "
+                        "can be defined. Please unify the versions or reduce the specs to use "
+                        "`spack develop`"
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]: This last sentence doesn't scan easily for me -- is the second suggestion supposed to be "or do not `spack develop` {spec.name}"?

File: lib/spack/spack/cmd/develop.py (hunk 7 above)
Comment (becker33) [reply lib/spack/spack/cmd/develop.py]: Which test was the one failing? I'm hesitant to just ignore this since it suggests something wrong with how we handle environments more broadly.

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/develop.py (hunk 11 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: Well no, but that is fine too. It's supposed to read, make the versions the same or delete specs until you only have the same version to use spack develop. I suppose that second one is redundant.

File: lib/spack/spack/cmd/develop.py (hunk 7 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: `test_develop_update_spec`

Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: With the current PR if I move this section back inside the `write_transaction` the following tests fail:

``` console
======================================================= short test summary info =======================================================
FAILED lib/spack/spack/test/cmd/develop.py::TestDevelop::test_develop_update_spec[mock_archive0] - AssertionError: assert 'mpich@=1.0' == 'mpich@=2.0'
FAILED lib/spack/spack/test/cmd/develop.py::test_recursive[mock_archive0] - AssertionError: assert 'indirect-mpich' in {'direct-mpich': {'path': 'direct-mpich', 'spec': 'direct-mpich@1.0/ldxba4og2st3nqkukqp...
```

Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: They pass if I check for the dev_entry outside of the `write_transaction`.

File: lib/spack/spack/cmd/develop.py (hunk 11 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: @becker33  I've updated the error message:
``` console
$ spack develop zlib-ng
==> Error: zlib-ng: has multiple concrete instances in the graph that can't be satisified by a single develop spec. To use `spack develop` ensure one of the following:
 a) zlib-ng nodes can satisfy the same develop spec (minimally this means they all share the same version)
 b) Provide a concrete develop spec (zlib-ng@[version]) to clearly indicate what should be developed
```

File: lib/spack/spack/cmd/develop.py (hunk 1 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: I think I'll mark this as resolved for now. I don't really know how to test this. I haven't hit an issue with it, but if someone wants to come up with some potential tests then I will be happy to run them.

File: lib/spack/spack/cmd/develop.py (hunk 3 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: @haampie will you weigh back in on this? I'd like to get this merged.

File: lib/spack/spack/cmd/develop.py [hunk 12]
```diff
@@ -83,86 +78,156 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
-        for name, entry in env.dev_specs.items():
-            path = entry.get("path", name)
-            abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
-
-            if os.path.exists(abspath):
-                msg = "Skipping developer download of %s" % entry["spec"]
-                msg += " because its path already exists."
-                tty.msg(msg)
-                continue
+def assure_concrete_spec(env: spack.environment.Environment, spec: spack.spec.Spec):
```
Comment (psakievich) [inline lib/spack/spack/cmd/develop.py:84]:
```suggestion
def assure_concrete_spec(env: "spack.environment.Environment", spec: spack.spec.Spec):
```

File: lib/spack/spack/cmd/develop.py [hunk 13]
```diff
@@ -38,28 +40,22 @@ def setup_parser(subparser):
         "--clone",
         action="store_true",
         dest="clone",
-        default=None,
+        default=True,
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py:42]: I don't understand this change -- this seems to suggest in the default case we clone regardless of whether the path is already cloned?

File: lib/spack/spack/cmd/develop.py [hunk 14]
```diff
@@ -83,86 +79,154 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+            raise SpackError("No spec provided to spack develop command")
+        else:
+            for name, entry in env.dev_specs.items():
+                path = entry.get("path", name)
+                abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
+                # Both old syntax `spack develop pkg@x` and new syntax `spack develop pkg@=x`
+                # are currently supported.
+                spec = spack.spec.parse_with_version_concrete(entry["spec"])
+                yield spec, abspath
+    else:
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]:
```suggestion
    if not args.spec:
        if args.clone is False:
            raise SpackError("No spec provided to spack develop command")
        for name, entry in env.dev_specs.items():
            path = entry.get("path", name)
            abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
            # Both old syntax `spack develop pkg@x` and new syntax `spack develop pkg@=x`
            # are currently supported.
            spec = spack.spec.parse_with_version_concrete(entry["spec"])
            yield spec, abspath
        return
```
This is purely cosmetic, to reduce the amount of indentation needed and make the code more legible. With this change then the entire rest of the function can be dedented by 4 spaces.

File: lib/spack/spack/cmd/develop.py [hunk 15]
```diff
@@ -83,86 +79,154 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+        specs = spack.cmd.parse_specs(args.spec)
+        if (args.path or args.build_directory) and len(specs) > 1:
+            raise SpackError(
+                "spack develop requires at most one named spec when using the --path or"
+                " --build-directory arguments"
+            )
+        else:
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]: Similarly, can we remove this `else` statement since the `if` block exits and dedent the entire else case?

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/develop.py (hunk 13 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py:42]: This is suggesting we call the `_clone` method and attempt to clone. Additional checks are there to explain why we skip cloning.  It's been a while but IIRC the logic was embedded off state of `None` `True` or `False` before.  But for a recursive clone we need better inferences that I moved inside a method.

**Comment by psakievich** [comment]
@spackbot fix style

File: lib/spack/spack/cmd/develop.py [hunk 16]
```diff
@@ -83,86 +79,154 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
...
+def assure_concrete_spec(env: spack.environment.Environment, spec: spack.spec.Spec):
+    version = spec.versions.concrete_range_as_version
+    if not version:
+        # first check environment for a matching concrete spec
+        matching_specs = env.all_matching_specs(spec)
+        if matching_specs:
+            version = matching_specs[0].version
+            test_spec = spack.spec.Spec(f"{spec}@{version}")
+            for m_spec in matching_specs:
+                if not test_spec.satisfies(m_spec):
```
Comment (becker33) [inline lib/spack/spack/cmd/develop.py]: I think this needs to be
```suggestion
                if not test_spec.intersects(m_spec):
```

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/develop.py (hunk 16 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py]: The order was wrong here. we should be checking that all the matching specs in the environment satisfy the test spec which was constructed from the name and version of the first matching spec.

**Comment by psakievich** [comment]
@spackbot fix style

File: lib/spack/spack/cmd/develop.py (hunk 13 above)
Comment (psakievich) [reply lib/spack/spack/cmd/develop.py:42]: Need to change the help messages.

Comment (psakievich) [reply lib/spack/spack/cmd/develop.py:42]: @becker33 updated this one

**Review by becker33 — DISMISSED** [review]

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
@spackbot fix style

**Review by becker33 — DISMISSED** [review]

**Review by becker33 — APPROVED** [review]
